File: lib/blocks/edge.c
#include "edge.h"

void block_Edge(struct Edge* self)
{
    bool signal = self->signal;
    bool reset  = self->reset;
    bool rising = self->rising;
    bool last   = self->last;

    if (reset)
    {
        struct Edge init = {};
        *self = init;
    }
    else if (rising)
    {
        self->last = signal;
        self->output = !last && signal;
    }
    else
    {
        self->last = signal;
        self->output = last && !signal;
    }
}

File: modules/edge.m
// -----------------------------------------------------------------------------
// Copyright 25 June 2023
// Author: Preston Langford
// This module debounces a boolean signal
// -----------------------------------------------------------------------------

#include "time.h"
#include "Edge.h"

impl Edge
{
    in bool reset;
    in bool input;
    in bool rising;

    var bool on = and(
        rising,
        input,
        not(state)
    );

    var bool off = and(
        not(rising),
        not(input),
        state
    );

    var bool state = and(input,not(reset));

    out bool output = and(
        or(on,off),
        not(reset)
    );
};
